aip/0162.md

```diff
@@ -0,0 +1,312 @@
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // A new revision is committed whenever the book is changed in any way.
+  string revision_id = 5 [
+    (google.api.field_behavior) = IMMUTABLE,
+    (google.api.field_behavior) = OUTPUT_ONLY];
+
+  // The timestamp that the revision was created.
+  google.protobuf.Timestamp revision_create_time = 6
+    [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The resource **must** contain a `revision_id` field, which **should** be a
+  string and contain a short, automatically-generated hash. A good rule of
+  thumb is the last eight characters of a UUID4.
+  - The `revision_id` field **must** document when new revisions are created
+    (see [committing revisions](#committing-revisions) below).
+- The resource **must** contain a `revision_create_time` field, which
+  **should** be a `google.protobuf.Timestamp` (see [AIP-142][]).
+
+### Referencing revisions
+
+When it is necessary to refer to a specific revision of a resource, APIs
+**must** use the following syntax: `{resource_name}@{revision_id}`. For
+example:
+
+    publishers/123/books/les-miserables@c7cfa2a8
+
+APIs **should** generally accept a resource reference at a particular revision
+in any place where they ordinarily accept the resource name. However, they
+**must not** accept a resource in situations that mutate the resource, and
+should error with `INVALID_ARGUMENT` if one is given.
+
+**Important:** APIs **must not** require a revision ID, and **must** default to
+the current revision if one is not provided, except in methods specifically
+dealing with the revision history (such as rollback) where failing to require
+it would not make sense (or lead to dangerous mistakes).
+
+### Getting a revision
+
+APIs implementing resource revisions **should** accept a resource name with a
+revision ID in the standard `Get` method ([AIP-131][]):
+
+```proto
+message GetBookRequest {
+  // The name of the book.
+  //   Example: publishers/123/books/les-miserables
+  //
+  // In order to retrieve a previous revision of the book, also provide
+  // the revision ID.
+  //   Example: publishers/123/books/les-miserables@c7cfa2a8
+  string name = 1;
+}
+```
+
+If the user passes a revision ID that does not exist, the API **must** fail
+with a `NOT_FOUND` error.
+
+APIs **must** return a `name` value corresponding to what the user sent. If the
+user sent a name with no revision ID, the returned `name` string **must not**
+include the revision ID either. Similarly, if the user sent a name with a
+revision ID, the returned `name` string **must** explicitly include it.
+
+APIs **should** allow users to specify a `@` with no revision ID to indicate
+that the user wants the latest revision, but for the revision ID to be
+explicitly included in the `name` field in the response:
+
+    publishers/123/books/les-miserables@
```

Comment:
I think that that this should be an error because it can cause bugs to lead to odd behavior-- For example the code
${NAME}@${REVISION} would succeed even if revision is unset.

it's not clear to me that this is better than having a sentinel value or just dropping the "@" all together and erroring in this case.

Reply:
Convinced.